feat: sync checkin configuration from config file

halfbyte opened the pull request:
Fixes #496

- CLI feature specs are a bin anemic, but the gist of the logic is in the sync service which is tested better
- Not sure if `checkin` should be `check_in` everywhere. I took inspiration from @subzero10's PRs on JS and Laravel where he consistently used `checkin` but both the API and the existing ruby code (for the actual check_in process) use the two word form. I'm happy to do the refactoring if we think that internal consistency is more important
- I think there's a good refactoring of the sync process possible that would reduce the amount of requests but I reused @subzero10's idea of caching the list of retrieved checkins to reduce the actual requests instead of the method calls.
- Also, not sure if this should be `honeybadger checkins sync` or `honeybadger sync checkins` instead and I didn't yet read the docs on how to build subcommands with Thor.


## TODO

- [ ] Remove `project_id` from config
- [ ] Modify check-in validation to require `slug` to be present
- [ ] Modify check-in validation to make `name` optional (make sure to test this because it's not clear in the API doc whether it's optional)
- [ ] To synchronize check-ins, fetch the project id to be used for the synchronization (see https://github.com/honeybadger-io/honeybadger/pull/4368)
- [ ] Use `slug` instead of `name` to check-in, by [constructing the check-in url](https://docs.honeybadger.io/guides/check-ins/#slugs) with API key and slug

subzero10 commented:
> I'm happy to do the refactoring if we think that internal consistency is more important

Consistency is important! The `checkin` vs `check_in` (or `checkIn`) inconsistency was already there when I started the PHP implementation and I (mistakenly) didn't think about correcting it. I just followed what was already there and then applied the same naming convention to the JavaScript implementation. I will rename all code from `checkin` to `checkIn` (we use camelCase in PHP and JavaScript) and apply similar convention to filenames (either `check-in` or `check_in`).

> Also, not sure if this should be honeybadger checkins sync or honeybadger sync checkins instead and I didn't yet read the docs on how to build subcommands with Thor.

I had the same dilemma. I ended up following shalvah's [suggestion](https://github.com/honeybadger-io/honeybadger-laravel/pull/111#issuecomment-1748869809) on this structure: `honeybadger DOMAIN ACTION`

halfbyte commented:
@subzero10 I've renamed everything, praised be ye global rename feature of the editors. I've also added the tabular output of the sync results which was missing. I'm planning on doing a couple of manual tests this afternoon to be sure I am not fooled by all the mocking and stubbing. 

I think you mentioned changing the signature of the actual check_in method so that it allows names instead of the id, but I'd love to do that in a separate PR if that's okay, this is big enough as is.

subzero10 commented:
> I think you mentioned changing the signature of the actual check_in method so that it allows names instead of the id, but I'd love to do that in a separate PR if that's okay, this is big enough as is.

OK, that works! Btw, it should allow both name or id (i.e it should not break any existing behavior).

subzero10 commented on lib/honeybadger/check_in.rb:
```diff
@@ -0,0 +1,87 @@
+      self.cron_timezone == other.cron_timezone
+    end
+
+    def to_json
```
Is there any chance you are going to be sending `nil` values for the optional properties? If that's the case, they should be changed to empty strings.
(@stympy correct me if I'm wrong)

stympy replied:
They should be nils, not empty strings.

subzero10 replied:
I will test the empty strings! If I understood well from our [conversation last time](https://honeybadger.slack.com/archives/C024F9JHC/p1696003452553299?thread_ts=1695989539.033289&cid=C024F9JHC), sending `nil` resulted in errors.

That's what led me to ignore optional values if they are not set (i.e. don't send them at all), but this creates an issues:
- If an optional field, such as `slug` is set, and now we want to unset it, it won't work because we won't be sending the `nil` value.

I think our best course of action here is to create constants of the defaults on the server and use them instead of `nil` (or empty strings).

stympy replied:
Thanks for the reminder! Yes, sending nulls for fields that have a no-null constraint won't work very well. :) I'm agreed on not sending optional values that aren't set.

> If an optional field, such as slug is set, and now we want to unset it, it won't work because we won't be sending the nil value.

We have special treatment for slug on the server-side... we nullify empty strings for that field, so you can send an empty string to unset the slug.

> I think our best course of action here is to create constants of the defaults on the server and use them instead of nil (or empty strings).

This sounds reasonable to me.

subzero10 replied:
The default values:
- `slug` -> `null` or `''`
- `grace_period` -> `00:00:00` or `''`

halfbyte replied:
Okay, so what I haven't really gathered from your conversation is if I should implement this now in this PR or if changes to handle this in the API are under way?

halfbyte replied:
Nevermind: I implemented roughly what's implemented in the PHP implementation but I wonder if I should be more selective with the payload (as in: only add the elements to the payload relevant for the `schedule_type`)

stympy replied:
> I wonder if I should be more selective with the payload (as in: only add the elements to the payload relevant for the schedule_type)

That would be preferred

stympy replied:
It might be best to send all the keys... I'm thinking of the scenario where someone changes from the simple schedule to the cron schedule or vice-versa. In those cases it could make sense to send them all in order to reset the now-unused values on the server side.

subzero10 replied:
> It might be best to send all the keys... I'm thinking of the scenario where someone changes from the simple schedule to the cron schedule or vice-versa.

Currently I'm not sending all the keys, only because I think it's not necessary (that's what I understood from looking at the backend).
But yeah, we could do some testing. @halfbyte can you please try changing a check-in from simple to cron and vice-verca?

stympy replied:
> Currently I'm not sending all the keys, only because I think it's not necessary (that's what I understood from looking at the backend).

True the backend _should_ be taking care of it. :)

halfbyte replied:
From what I can see, the backend handles missing keys just fine and also seems to reset fields for the "inactive" schedule type properly. If I change a simple check in to type "cron" with a proper `cron_schedule` and then try to change it back to "simple" without also setting `report_period` the server throws an error.

I'll adapt my code to closer follow the PHP implementation, I think?

subzero10 replied:
> From what I can see, the backend handles missing keys just fine and also seems to reset fields for the "inactive" schedule type properly. If I change a simple check in to type "cron" with a proper cron_schedule and then try to change it back to "simple" without also setting report_period the server throws an error.

This is the correct behavior though, right?
I would expect that you would receive an error before even reaching the server, when the check-ins are loaded from the config file and [validated](https://github.com/honeybadger-io/honeybadger-ruby/pull/508/files#diff-2306255ad6db4971c5877c2f3b98218488fca0fb02f6b95895df8b65a604c5e4R59).

> I'll adapt my code to closer follow the PHP implementation, I think?

What would you change here? I thought that this part follows the PHP implementation as it is now.

subzero10 commented on lib/honeybadger/backend/server.rb:
```diff
@@ -48,8 +49,103 @@ def check_in(id)
+      def update_check_in(project_id, id, check_in_config)
+        response = Response.new(@http.put("/v2/projects/#{project_id}/check_ins/#{id}", check_in_config.to_json, personal_auth_headers))
+        if response.success?
+          return CheckIn.from_remote(project_id, JSON.parse(response.body))
```
Did you test this route? If I remember well, it does not return a body; only a HTTP 204 (no content).

subzero10 commented on lib/honeybadger/config_sync_service.rb:
```diff
@@ -0,0 +1,74 @@
+      created_or_updated = sync_existing_checkins(checkins)
+      removed = sync_removed_checkins(checkins)
+
+      return (created_or_updated + removed).uniq
```
You shouldn't need to call `uniq` here, right?

subzero10 commented on lib/honeybadger/config_sync_service.rb:
```diff
@@ -0,0 +1,74 @@
+      checkin_configs = @config.get(:checkins) || []
+      return [] if checkin_configs.empty?
+
+      checkins = checkin_configs.map do |cfg|
```
I think there should be a check here whether there are any checkIns with the same names (under the same project id) and raise an error in that case.

halfbyte replied:
I looked at the laravel (or rather php) code for this and I don't see a similar check? If I'm missing it could you point to it?

Also, I wonder, if this is something that should be enforced if it would be much easier to enforce it server side?

(I agree that enforcing unique names makes sense if we want to allow people to use the name for running the check in)

stympy replied:
Unique names per project are enforced on the server side.

subzero10 replied:
@halfbyte [here](https://github.com/honeybadger-io/honeybadger-php/blob/f7d3a5902ce370e7fae22f084c5b9162f86035ab/src/CheckInsManager.php#L59) it is. You might have missed this because I only added the check last week.

> Unique names per project are enforced on the server side.

That's good! The issue for the client side is that whichever check-in comes last will overwrite the previous (at least that's what I experienced with PHP and JS) - so no error, just unexpected results.

stympy replied:
Oops - I lied... I was thinking of slugs. We don't currently require unique check-in names. I'll make that change. :)

subzero10 replied:
@stympy Are you making them unique globally or per project?
Because in our examples and documentation, we have check-in names such as _Hourly Notifications_. That's not a contestant for a globally unique name 😅. Besides, check-in ids are globally unique, so names could be unique per project, right?

stympy replied:
The uniqueness check for the name will be per-project.

subzero10 replied:
@halfbyte Are you planning to do implement this check here? I think it will help users with unexpected behavior. Consider the following:
- 2 _new_ checkins => `[ { name: 'my check-in', scheduleType: 'simple' }, { name: 'my check-in', scheduleType: 'cron' } ]`
- The sync command will go through the first check-in and call the backend to find out whether this is a new check-in or it already exists. Since it does not exist, it will issue a create new check-in request.
- For the second check-in, the backend (or the local in-memory cache) will return the same check-in and issue an update request.
- So, no error will be reported, but the check-ins defined in the config won't match what's on Honeybadger.

halfbyte replied:
@subzero10 I've added a simple check similar to the PHP one.

subzero10 commented on lib/honeybadger/config_sync_service.rb:
```diff
@@ -0,0 +1,74 @@
+          local_project_checkins.find{|c| c.id == pc.id || c.name == pc.name }
+        end
+        to_remove.each do |ch|
+          if ch.id.nil?
```
This shouldn't happen:
- If a check-in is to be removed, it means that it is available remotely (i.e. it came from the HB API) but does not exist in the local config array
- Since it came from the HB API, it must have an id

subzero10 reviewed (commented):
This looks nice! I left a few comments!

stympy commented on lib/honeybadger/check_in.rb:
```diff
@@ -0,0 +1,87 @@
+  class InvalidCheckinConfig < StandardError; end
+  class CheckInSyncError < StandardError; end
+  
+  class CheckIn
```
This seems like a lot of code for what is essentially shuffling hashes back and forth. What about something like this?

```ruby
# frozen_string_literal: true

class CheckIn
  ATTRIBUTES = %w[name slug schedule_type report_period grace_period cron_schedule cron_timezone]

  attr_reader :project_id, :data

  def self.from_config(atts)
    new(atts["project_id"], atts["id"], atts)
  end

  def self.from_remote(project_id, atts)
    new(project_id, atts["id"], atts)
  end

  def initialize(project_id, id: nil, atts: nil)
    @project_id = project_id
    @id = id
    @data = atts.transform_keys(&:to_s).slice(*ATTRIBUTES) if atts.is_a?(Hash)
  end

  def ==(other)
    data == other.data
  end

  def to_json
    data.to_json
  end

  ATTRIBUTES.each do |meth|
    define_method meth do
      data[meth]
    end
  end

  ...
end
```

halfbyte replied:
@stympy This was a conscious decision on my part as I usually value the explicitness over brevity but I'll happily go with your approach if that's more of your liking.

stympy replied:
Got it! Yeah, I think we'd prefer brevity. Thanks for your flexibility!

stympy commented on lib/honeybadger/config/defaults.rb:
```diff
@@ -328,6 +333,11 @@ class Boolean; end
         default: true,
         type: Boolean
+      },
+      :'checkins' => {
```
```suggestion
      checkins: {
```

stympy commented on lib/honeybadger/cli/main.rb:
```diff
@@ -117,26 +118,27 @@ def exec(*args)
-
-      desc 'heroku SUBCOMMAND ...ARGS', 'Manage Honeybadger on Heroku'
+      
+      desc 'heroku SUBCOMMAND ...ARGS', 'Manage -- Honeybadger on Heroku'
```
```suggestion
      desc 'heroku SUBCOMMAND ...ARGS', 'Manage Honeybadger on Heroku'
```

stympy reviewed (approved):
LGTM. I'd like @joshuap to take a look, too.

subzero10 commented on lib/honeybadger/backend/server.rb:
```diff
@@ -48,8 +49,103 @@ def check_in(id)
       private
 
+      def personal_auth_headers
+        {"Authorization" => "#{@personal_auth_token}:"}
```
Is this how HTTP Basic Auth is done?

subzero10 replied:
Doesn't this need to be base64 encoded?

halfbyte replied:
Yeah, no, this looks dubious on first glance. I'll test and correct.

halfbyte replied:
Ooof, I completely mocked myself here. As in: I never really tested this against the real thing and had no good tests for the authentication. Sorry for that. I also realised that I needed to refactor the server access a bit, as the new API sits on `app.honeybadger.io` instead of `api.honeybadger.io`.  I have now tested this successfully and cleaned up the code a bit as well.

@subzero10 Please take another look.

subzero10 commented on lib/honeybadger/cli/check_ins.rb:
```diff
@@ -0,0 +1,55 @@
+      extend Forwardable
+
+      class_option :personal_auth_token, required: false, type: :string, desc: "personal auth token for API access"
+      class_option :environment,     required: false, aliases: [:'-e', :'-env'], type: :string, desc: 'Environment this command is being executed in (i.e. "production", "staging")'
```
What would be the use case to call the sync CLI command per environment?

I understand that when executing the check-in api call, we could allow passing the environment variable as a parameter in order to dictate that the check-in api call should be made only when the environment matches (example [here](https://docs.honeybadger.io/lib/php/integration/laravel/#after-a-scheduled-command)).

However, this is the sync CLI command and ideally it should only run on the production environment, right?

halfbyte replied:
@subzero10 The others may prove me wrong, but I think this could be important as rails technically does allow for multiple "production" environments, as it is not restricted to "development", "test" and "production" as other frameworks. So if the honeybadger config is actually set in a rails initializer, it could be environment dependent.

So, for example, the sync cli command could also run for a "staging" environment" and that could load/should load different checks.

subzero10 replied:
> So, for example, the sync cli command could also run for a "staging" environment" and that could load/should load different checks.

I see, but I think that wouldn't work in this case. As far as I know, the backend does not support check-ins per environment.
However, I'm wondering how would you know which config file to read if no environment is passed as an argument. Or in other words, which one is the "production" environment?

halfbyte replied:
Maybe @stympy or @joshuap could chime in here. My thinking was that a honeybadger initializer in a rails app could do checks like `if Rails.env.staging?` and do different things depending on the loaded environment. I would assume different "production" environments would use different Honeybadger projects altogether and thus also would use different checkins. 

This might all be a bit theoretical but the fact that config *can* be set from within Rails configuration files/initializers makes me quite convinced the environment is important information the user should be able to pass in here.

subzero10 replied:
> I would assume different "production" environments would use different Honeybadger projects altogether and thus also would use different checkins.

That is a fair point that I didn't consider 😄. This isn't a blocker for me. We can proceed as it is and we may come back to it in the future if it causes confusion.

joshuap replied:
I see you can set the `project_id` per check-in in the `check_ins` config, so maybe that's enough—people can sync check-ins for one or many projects already?

subzero10 replied:
> I think the best course here is to build the simplest thing (don't consider environments at all in the Rails integration), and then we could tackle an official solution later if people want it. In the meantime, it would be nice if we could suggest a manual workaround for anyone who does want to use multiple Honeybadger projects in the same Rails app.

I agree with this approach and that's what prompted me to ask about the _environment_ option in the first place 😄.

> I see you can set the project_id per check-in in the check_ins config, so maybe that's enough—people can sync check-ins for one or many projects already?

Indeed, that is a consequence of having to supply a `project_id`, since we can't use the project API key at the moment.

subzero10 reviewed (commented):
I think this works as it is, but I left a couple of comments.

halfbyte commented:
@subzero10 PTAL

subzero10 reviewed (approved):
This looks good to me!

subzero10 commented:
@halfbyte Reminder that we also have to do the following when this PR is merged:
- change the signature of the actual check_in method so that it allows names or ids

halfbyte commented:
>     * change the signature of the actual check_in method so that it allows names or ids

See #509

Should I wait for @joshuap's review? I think so?

stympy commented:
> Should I wait for @joshuap's review? I think so?

Yup :)

joshuap commented on lib/honeybadger/check_in.rb:
```diff
@@ -0,0 +1,84 @@
+    private 
+
+    def blank?(str)
+      str.nil? || str == "" 
```
This would also catch whitespace:

```suggestion
      str.nil? || str.strip == "" 
```

joshuap commented on lib/honeybadger/config/defaults.rb:
```diff
@@ -141,6 +146,11 @@ class Boolean; end
         type: String
       },
+      :'connection.app_host' => {
+        description: 'The host to use when interacting with the new API.',
```
```suggestion
        description: 'The host to use when interacting with the REST API.',
```

joshuap commented on lib/honeybadger/config_sync_service.rb:
```diff
@@ -0,0 +1,82 @@
+    def check_unique_names(check_ins)
+      names = check_ins.map(&:name)
+      dupes = names.find_all {|n| names.count(n) > 1}.uniq
+      raise Honeybadger::InvalidCheckinConfig.new("Check Ins need to have unique names. #{dupes.join(", ")} used multiple times.") if dupes.length > 0
```
```suggestion
      raise Honeybadger::InvalidCheckinConfig.new("Check-ins need to have unique names. #{dupes.join(", ")} used multiple times.") if dupes.length > 0
```

joshuap commented on lib/honeybadger/config_sync_service.rb:
```diff
@@ -0,0 +1,82 @@
+    end
+
+    def sync_removed_check_ins(check_ins)
+      return [] if check_ins.nil? || check_ins.empty?
```
@subzero10 @halfbyte I like this guard and think it's essential that we don't delete people's existing check-ins if they run `honeybadger check_ins sync` without having `check_ins` config. That said, do we also provide any kind of warning for people who want to try this feature and have a lot of existing check-ins? For example, if I have an existing Honeybadger project with check-ins, and I try to use this feature, my first assumption will not be "I need to put _all_ of my check-ins into my local config before running this command, otherwise they will be deleted."

Obviously we should highlight this behavior in the docs, but I'm wondering if we can do anything else to prevent accidents. Some people have hundreds of check-ins. 😬

subzero10 replied:
> Some people have hundreds of check-ins.

This feature is exactly for these people! Imagine that clicks they had to go through until they created those checkins.

Here are some ideas (on top of the warning in the documentation):
- We could create a `honeybadger check-ins pull` command that you run once and populates your config file with all your check-ins.
- We could implement an `inactive` state for check-ins and mark check-ins as inactive instead of deleting them. This would allow people to easily _rollback_ if they accidentally disable check-ins.

joshuap replied:
> Another idea would be to prompt for confirmation in the CLI, but I wouldn't recommend it since we encourage people to use the command as part of their deployment pipelines.

One option is to prompt by default, but have a flag to skip the prompt for deployments.

joshuap replied:
I also like the `honeybadger check-ins pull` idea — not as a solution for this problem, but it would be handy when migrating to this (we could consider adding it later).

subzero10 replied:
Sounds good! So, to sum up:
- Adopt the CLI to prompt for confirmation before going ahead.
- Create a CLI option to skip the prompt, to be used with CI environments.
- When writing the documentation page, make sure to add a callout saying that using the CLI will overwrite any existing check-ins not defined in the configuration file.


I suggest that we decide on the prompt, so that we can apply it on all libraries, how about this one:
```
Warning: Proceeding will replace all your current check-ins with the ones defined in the configuration file. Do you want to continue? (y/n)
```
What do you think?

halfbyte replied:
@subzero10 @joshuap I've added that prompt (took the liberty to add a line break) and added a `--yes` command line switch to override the confirmation. Not sure that's the best switch for this but it seems to be somewhat common for these kinda things, but let me know if you want to have it changed.

joshuap replied:
apt uses `-y`/`--yes` when installing packages. I think that works.

@stympy @rabidpraxis @roelbondoc @subzero10 does that make sense to you?

joshuap commented on lib/honeybadger/util/http.rb:
```diff
@@ -37,8 +37,8 @@ def initialize(config)
-      def get(endpoint)
-        response = http_connection.get(endpoint)
+      def get(endpoint, headers = nil)
+        response = http_connection.get(endpoint, http_headers(headers))
```
This get method is used when performing a check in here:

https://github.com/honeybadger-io/honeybadger-ruby/blob/aabc35a1612c420881aa2c77bd3f624fd3af8c89/lib/honeybadger/backend/server.rb#L41-L50

This change means that these headers will be included with those requests (which is a change in existing behavior):

https://github.com/honeybadger-io/honeybadger-ruby/blob/aabc35a1612c420881aa2c77bd3f624fd3af8c89/lib/honeybadger/util/http.rb#L17-L22

and:

https://github.com/honeybadger-io/honeybadger-ruby/blob/aabc35a1612c420881aa2c77bd3f624fd3af8c89/lib/honeybadger/util/http.rb#L63

This doesn't seem desirable to me; for example, the [check-ins API](https://docs.honeybadger.io/api/reporting-check-ins/) does not use JSON, there is no compression happening, etc.

Thoughts @stympy @subzero10?

subzero10 replied:
🤔 I think this was added to be able to make HTTP GET for the `app_http.rb` class, right @halfbyte ? If that's the case, can we:
- create a `base.rb` class with some shared functionality (i.e. `setup_http_connection`)
- create `app_http.rb` and `api_http.rb` with different definitions for HTTP GET

halfbyte replied:
I'll just move my modified `#get` to `AppHTTP`, that should fix this. I didn't think of the consequences, really, thanks @joshuap for catching that.

joshuap replied:
I think that works for now. We might need to refactor the http client helpers if we add to them in the future.

joshuap reviewed (changes requested):
Looking good! I'm not _necessarily_ requesting changes here, but I have a few questions which may result in changes (and I made a few small code suggestions). :)

Another random thought I just had: I see we're using the `project_id` from the `check_ins` config everywhere. Have you considered defaulting to the project `api_key` if people exclude the `project_id`? We'd have to either invent a way to fetch the real `project_id` to use for substitution, or we'd have to support the project `api_key` in addition to the `project_id` in the REST API. I don't love either of those options, but I do like the experience of Honeybadger already knowing what my project is, since it already has my project API key (and most people will use this with a single project). 🤔

subzero10 commented:
> Another random thought I just had: I see we're using the project_id from the check_ins config everywhere. 

As you said, the `project_id` is necessary today, because we have to specify which project we are creating check-ins for.
In my opinion, the ideal solution would be to be able to use the _project api key_ and skip the _project_id_ entirely, maybe even the _personal auth token_.

> Have you considered defaulting to the project api_key if people exclude the project_id?

Is this possible with the current API?

> We'd have to either invent a way to fetch the real project_id to use for substitution

I think I like this idea.
We could:
- have a route that that takes a project api key and returns a project id (route would be authenticated with the personal auth token)
- drop the `project_id` field entirely from each check-in

> (and most people will use this with a single project)

True. If we come to a solution with what you suggest above, we'll love this multi-project check-ins possibility, but that's something I'm perfectly fine with.

joshuap commented:
> @subzero10 @joshuap I think I followed your conversation about the `project_id` thing but I am not super sure if I can draw any conclusions regarding this PR from it right now? :)

I don't think there are any conclusion to draw just yet, we just need to hash this out before moving forward! 😄

joshuap commented on lib/honeybadger/cli/check_ins.rb:
```diff
@@ -0,0 +1,58 @@
+
+      class_option :personal_auth_token, required: false, type: :string, desc: "personal auth token for API access"
+      class_option :skip_rails_load, required: false, type: :boolean, desc: 'Flag to skip rails initialization'
+      class_option :yes, required: false, type: :boolean, desc: 'Flag to skip confirmation'
```
```suggestion
      class_option :yes, required: false, aliases: :'-y', type: :boolean, desc: 'Flag to skip confirmation'
```

joshuap commented:
Hey all, an update on this PR:

We're going to hold off on merging this for now, and I may close it at some point—but we may pick it back up next year. We just have other priorities to focus on for the moment.

In the meantime, I would like to finish the discussions we are having here so that we'll be in a good place to move forward if/when come back to it. So let's continue discussing the details on the few unresolved threads above, but otherwise we can shift focus to other projects.

(@stympy and @subzero10 — please ignore the review requests above, I was just resetting the approvals.)

joshuap commented:
> > I think this is OK for syncing things, and is a nicer default experience. Since we already have the per-check-in project_id option in this PR, is there any harm in leaving it as an override if some people want to sync check-ins to other projects?
> 
> Since we now have the ability to map a project api key to a project id ([honeybadger-io/honeybadger#4368](https://github.com/honeybadger-io/honeybadger/pull/4368)), I would vote to remove entirely `project_id` from config. If we have requests to support multiple projects, we can consider bringing it back in the future.

This works for me. 👍 

> If we were to change something in the PR, it would be:
> 
> * remove `project_id` from config
> * before synchronizing check-ins, first fetch the project id to be used for the synchronization
> * modify validation to require `slug` to be present

See my https://github.com/honeybadger-io/honeybadger-ruby/issues/509#issuecomment-1837041129 for one more suggestion—can we use the slug instead of the name as the sync identifier when the check-in ID isn't available? This would reduce the check-in name to just a label again, which I think is better—people might want to rename them without affecting sync. (Let me know if I'm missing something here.)

> > I'm less certain about performing check-ins. I'd really like to keep it so that we don't need to perform an extra request when actually performing a check in, as discussed in #509. Is that relevant here?
> 
> For performing check-ins, we will use api key and slug to construct the check-in url, which can be done separately for #509.

Yep, I think this a nicer experience—you don't need to worry about project_id at all as the user. 👍

subzero10 commented:
> See my https://github.com/honeybadger-io/honeybadger-ruby/issues/509#issuecomment-1837041129 for one more suggestion—can we use the slug instead of the name as the sync identifier when the check-in ID isn't available? This would reduce the check-in name to just a label again, which I think is better—people might want to rename them without affecting sync. (Let me know if I'm missing something here.)

No you are not missing anything, one of the reasons to make slug required is because we will use it to sync the check-ins as well. I just forgot to type it in the changes to be made! I will modify the PR description and add the complete list of remaining todo items.

halfbyte commented on lib/honeybadger/util/app_http.rb:
```diff
@@ -0,0 +1,44 @@
+
+      def http_headers(headers = nil)
+        {}.tap do |hash|
+          hash.merge!(HEADERS)
```
- Content-Type - Needed
- Content-Encoding - Works with the Rest API. Maybe it would be desirable to switch it off as it makes testing a tiny bit easier, on the other hand, you know, compression.
- Accept: Makes sense to me?
- User-Agend: Makes sense, I guess?

joshuap commented:
Closing this since we don't have the bandwidth currently. We'll keep the branch around and create a new PR if/when we resume this feature. Thanks everyon!
